### PR TITLE
feat(Dataset): Negative Rows for Classification.

### PR DESCRIPTION
**Pull request by szemyd**
-1 for negative labels (where there is no bug)
+1 for positive labels (where there was a bug)

### CODE REVIEW

File: extractors/split_utils.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import pandas as pd
+import numpy as np
+from sklearn.model_selection import train_test_split
+from typing import Tuple
+
+
+def add_negative_cases(df: pd.DataFrame) -> pd.DataFrame:
+    positive = df.copy()[["before"]]
```
Comment (itchingpixels) [inline extractors/split_utils.py:8]: not sure if we need to copy here, as we don't change any of its values. copying can be quite costly when we scale up the dataset, and we'll probably want to do that!

File: extractors/split_utils.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    print(f"| Adding labels to rows...")
+    negative["label"] = 0.0
+    negative = negative.rename(columns={"after": "source_code"})
+    positive["label"] = 1.0
+    positive = positive.rename(columns={"before": "source_code"})
+
+    print(f"| Combining and shuffling dataframe...")
+    df = pd.concat([positive, negative])
+
+    df = df.iloc[np.random.permutation(len(df))]
```
[the remark this answers is not in the document]
Comment (itchingpixels) [reply extractors/split_utils.py:20]: also we don't need this, as train_test_split automatically shuffles

File: extractors/split_utils.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+
+from sklearn.model_selection import train_test_split
+
+
+def split_data(df: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    test_ratio = 0.1
+    val_ratio = 0.1
+    train, val_test = train_test_split(df, test_size=test_ratio + val_ratio)
+    val, test = train_test_split(val_test, test_size=test_ratio)
```
Comment (itchingpixels) [inline extractors/split_utils.py:31]: here test_ratio is only 0.1, is that intentional? so our test dataset will be (0.1 + 0.1) * 0.1

File: run_classification.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    def compute_metrics(eval_pred):
+        logits, labels = eval_pred
+        predictions = np.argmax(logits, axis=-1)
+        return metric.compute(predictions=predictions, references=labels)
+
+    # training_args = TrainingArguments(
+    #     output_dir="test_trainer", evaluation_strategy="epoch", report_to="none"
+    # )
+    # metric = load_metric("accuracy")
+    metric = load_metric("f1", average="weighted")
```
Comment (itchingpixels) [inline run_classification.py:61]: I don't think it makes sense to use weighted f1 for binary classification - that only makes sense in a multi-class classification task

File: run_extract_embeddings_diff.py [hunk 5]
```diff
@@ -28,16 +30,14 @@ def extract_embeddings_from_pipeline(
...
     if config.force_cpu:
         print("| Setting device to cpu...")
         device = -1
 
-    print("| Running preprocessing...")
-    df = pd.read_json(
-        f"data/original/TSSB-3M/file-0.jsonl.gz", lines=True, compression="gzip"
-    )[: config.dataset_size]
+    df = load_data(config)
+    df = add_negative_cases(df)
```
Comment (itchingpixels) [inline run_extract_embeddings_diff.py:40]: I don't think we can use negative cases here, or?

File: extractors/pipeline.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
 class MyFeatureExtractionPipeline(FeatureExtractionPipeline):
     def preprocess(self, inputs, truncation=None) -> Dict[str, t.Tensor]:
+        print(self.tokenizer)
```
Comment (itchingpixels) [inline extractors/pipeline.py:8]:
```suggestion
```

File: extractors/split_utils.py (hunk 3 above)
Comment (szemyd) [reply extractors/split_utils.py:31]: You are right, this is an error I skipped while I was putting the pipeline together.

File: run_extract_embeddings_diff.py (hunk 5 above)
Comment (szemyd) [reply run_extract_embeddings_diff.py:40]: I have no idea how this ended up here. I was tired when I refractored this 🙈

File: environment.yml [hunk 7]
```diff
@@ -26,5 +26,6 @@ dependencies:
   - pyparsing
   - pip
   - rich
+  - sklearn
```
[the remark this answers is not in the document]
Comment (szemyd) [reply environment.yml:29]: 👍

File: extractors/split_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (szemyd) [reply extractors/split_utils.py:20]: Yes!

File: extractors/split_utils.py (hunk 1 above)
Comment (szemyd) [reply extractors/split_utils.py:8]: I thought the selector df[["before"]]  references the columns of the original but I see they return new dfs.
